votor/src/timer_manager/timers.rs

```diff
@@ -0,0 +1,156 @@
+/// timeouts for a window of slots.
+enum TimerState {
+    /// Waiting for the DELTA_TIMEOUT stage.
+    WaitDeltaTimeout {
+        /// The slots in the window.  Must not be empty.
+        window: Vec<Slot>,
+        /// Time when this stage will end.
+        timeout: Instant,
+    },
+    /// Waiting for the DELTA_BLOCK stage.
+    WaitDeltaBlock {
+        /// The slots in the window.  Must not be empty.
+        window: Vec<Slot>,
+        /// Time when this stage will end.
+        timeout: Instant,
+    },
+    /// The state machine is done.
+    Done,
+}
+
+impl TimerState {
+    /// Creates a new instance of the state machine.
+    ///
+    /// Also returns the next time the timer should fire.
+    fn new(slot: Slot) -> (Self, Instant) {
+        let mut window = (slot..last_of_consecutive_leader_slots(slot)).collect::<Vec<_>>();
+        assert!(!window.is_empty());
+        window.reverse();
+        let timeout = Instant::now() + DELTA_TIMEOUT;
+        (Self::WaitDeltaTimeout { window, timeout }, timeout)
+    }
+
+    /// Call to make progress on the state machine.
+    ///
+    /// Returns a potentially empty list of events that should be sent.
+    fn progress(&mut self) -> Vec<VotorEvent> {
+        let now = Instant::now();
+        match self {
+            Self::WaitDeltaTimeout { window, timeout } => {
+                if &now < timeout {
+                    return vec![];
+                }
+                let slot = *window.last().unwrap();
+                let timeout = now + DELTA_BLOCK;
+                *self = Self::WaitDeltaBlock {
+                    window: window.to_owned(),
+                    timeout,
+                };
+                vec![VotorEvent::TimeoutCrashedLeader(slot)]
+            }
+            Self::WaitDeltaBlock { window, timeout } => {
+                if &now < timeout {
+                    return vec![];
+                }
+
+                let mut events = Vec::with_capacity(2);
+                let slot = window.pop().unwrap();
+                events.push(VotorEvent::Timeout(slot));
+                if window.is_empty() {
+                    *self = Self::Done;
+                    return events;
+                }
+                events.push(VotorEvent::TimeoutCrashedLeader(slot));
+                *timeout = now + DELTA_BLOCK;
+                events
+            }
+            Self::Done => vec![],
+        }
+    }
+
+    /// When would this state machine next be able to make progress.
+    fn next_fire(&self) -> Option<Instant> {
+        match self {
+            Self::WaitDeltaTimeout { window: _, timeout } => Some(*timeout),
+            Self::WaitDeltaBlock { window: _, timeout } => Some(*timeout),
+            Self::Done => None,
+        }
+    }
+}
+
+/// Maintains all active timer states for windows of slots.
+pub(super) struct Timers {
+    /// Timers are indexed by slots.  It is possible to have multiple active
```

Comment:
I don't see the reason we need to have multiple active timers for a given slot, the first one firing should take care of any skip votes that need to be sent.